File: common/src/main/java/org/sandboxpowered/silica/registry/SilicaRegistry.java
package org.sandboxpowered.silica.registry;

import org.jetbrains.annotations.Nullable;
import org.sandboxpowered.api.content.Content;
import org.sandboxpowered.api.registry.Registry;
import org.sandboxpowered.api.util.Identity;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Stream;

public class SilicaRegistry<T extends Content<T>> implements Registry<T> {
    private final Identity registryId;
    private final Class<T> type;

    public Map<Identity, T> internalMap = new HashMap<>();

    public Map<Identity, SilicaRegistryEntry<T>> registryEntries = new HashMap<>();

    public SilicaRegistry(Identity registryId, Class<T> type) {
        this.registryId = registryId;
        this.type = type;
    }

    public <X extends Content<X>> Registry<X> cast() {
        return (Registry<X>) this;
    }

    @Override
    public Identity getIdentity(T val) {
        return val.getIdentity();
    }

    @Override
    public Entry<T> get(Identity identity) {
        return registryEntries.computeIfAbsent(identity, id -> new SilicaRegistryEntry<>(this, id));
    }

    @Override
    public Entry<T> register(T val) {
        internalMap.put(val.getIdentity(), val);
        return get(val.getIdentity());
    }

    @Override
    public Stream<T> stream() {
        return registryEntries.values().stream().filter(Entry::isPresent).map(Entry::get);
    }

    @Override
    public Collection<Identity> keys() {
        return registryEntries.keySet();
    }

    @Override
    public Class<T> getType() {
        return type;
    }

    @Override
    public Identity getIdentity() {
        return registryId;
    }

    public void clearCache() {
        registryEntries.forEach((id, aEntry) -> aEntry.clearCache());
    }

    public static class SilicaRegistryEntry<T extends Content<T>> implements Registry.Entry<T> {
        private final SilicaRegistry<T> registry;
        private final Identity target;
        private boolean hasCached;
        private T cachedValue;

        public SilicaRegistryEntry(SilicaRegistry<T> registry, Identity target) {
            this.registry = registry;
            this.target = target;
        }

        public void updateCache() {
            if (!hasCached) {
                cachedValue = registry.internalMap.get(target);
                hasCached = true;
            }
        }

        public void clearCache() {
            cachedValue = null;
            hasCached = false;
        }

        @Nullable
        public T getInternal() {
            updateCache();
            return cachedValue;
        }

        @Override
        public T get() {
            return getInternal();
        }

        @Override
        public Optional<T> getAsOptional() {
            return Optional.ofNullable(getInternal());
        }

        @Override
        public T orElse(T other) {
            T val = getInternal();
            return val == null ? other : val;
        }

        @Override
        public T orElseGet(Supplier<T> other) {
            T val = getInternal();
            return val == null ? other.get() : val;
        }

        @Override
        public boolean isPresent() {
            updateCache();
            return cachedValue != null;
        }

        @Override
        public boolean matches(T other) {
            T val = getInternal();
            return val == other;
        }

        @Override
        public void ifPresent(Consumer<T> tConsumer) {
            T val = getInternal();
            if (val != null)
                tConsumer.accept(val);
        }

        @Override
        public void ifPresentOrElse(Consumer<T> tConsumer, Runnable notPresent) {
            T val = getInternal();
            if (val != null)
                tConsumer.accept(val);
            else
                notPresent.run();
        }
    }
}

File: common/src/main/kotlin/org/sandboxpowered/silica/registry/SilicaRegistry.kt
package org.sandboxpowered.silica.registry

import org.sandboxpowered.silica.util.Identifier
import java.util.*
import java.util.function.Supplier

class SilicaRegistry<T : RegistryEntry<T>>(private val id: Identifier, override val type: Class<T>) : Registry<T> {
    var internalMap: MutableMap<Identifier, T> = HashMap()
    var registryEntries: MutableMap<Identifier, SilicaRegistryEntry<T>> = HashMap()
    var listeners: MutableList<(T) -> Unit> = ArrayList()

    fun <X : RegistryEntry<X>> cast(): Registry<X> = this as Registry<X>

    fun addListener(listener: (T) -> Unit) {
        listeners.add(listener)
    }

    fun register(t: T): T {
        internalMap[t.identifier] = t
        listeners.forEach { it(t) }
        return t
    }

    override fun iterator() = internalMap.values.iterator()

    override fun contains(id: Identifier) = internalMap.containsKey(id)

    override fun getId(element: T) = element.identifier

    override fun getUnsafe(id: Identifier) = internalMap[id]

    override fun get(id: Identifier) =
        registryEntries.computeIfAbsent(id) { SilicaRegistryEntry(this, it) }

    override fun stream() =
        registryEntries.values.stream().filter { obj: SilicaRegistryEntry<T> -> obj.isPresent }
            .map { obj: SilicaRegistryEntry<T> -> obj.get() }

    fun clearCache() {
        registryEntries.forEach { (_, entry: SilicaRegistryEntry<T>) -> entry.clearCache() }
    }

    class SilicaRegistryEntry<T : RegistryEntry<T>>(
        override val registry: SilicaRegistry<T>,
        override val id: Identifier
    ) : RegistryObject<T> {
        private var hasCached = false
        private var cachedValue: T? = null

        private fun updateCache() {
            if (!hasCached) {
                cachedValue = registry.internalMap[id]
                hasCached = true
            }
        }

        fun clearCache() {
            cachedValue = null
            hasCached = false
        }

        private val internal: T?
            get() {
                updateCache()
                return cachedValue
            }
        override val isPresent: Boolean
            get() {
                updateCache()
                return cachedValue != null
            }
        override val isEmpty: Boolean
            get() = !isPresent

        override fun get() = internal ?: throw NullPointerException()

        override fun asOptional() = Optional.ofNullable(internal)

        override fun or(supplier: RegistryObject<T>) = if (!this.isPresent) supplier else this

        override fun <X : Throwable> orElseThrow(supplier: Supplier<X>) = internal ?: throw supplier.get()

        override fun orElseGet(supplier: Supplier<T>) = internal ?: supplier.get()
    }
}